RO_optimization for frequency and amplitude

maxhant opened the pull request:
Two new routines to optimize readout pulse's amplitude and frequency using `sweeper`

It was tested with qblox sweepers: https://github.com/qiboteam/qibolab/tree/alvaro/qblox9_sweepers
with the following actions:
```yml
platform: qblox
runcard: /home/users/maxime.hantute/qibolab/src/qibolab/runcards/qw5q_gold_qblox.yml

# Instructions: Uncomment whatever you would like to run!

qubits: [0,1,2,3,4]


format: csv

actions:

  ro_frequency:
    frequency_width: 2.e+6
    frequency_step: .1e+6
    nshots: 2000

  ro_amplitude:
    amplitude_factor_min: 0.1
    amplitude_factor_max: 0.3
    amplitude_factor_step: 0.01
    nshots: 2000
```

Here are samples of the results
![image](https://user-images.githubusercontent.com/52139175/234303643-d7277158-7db9-4213-b73c-567026c745e7.png)
![image](https://user-images.githubusercontent.com/52139175/234303996-04339611-d3aa-467e-a4cf-4ab7aac36153.png)


Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

andrea-pasquale commented:
@maxhant I tried to run it and I got the following error
```
AttributeError: 'list' object has no attribute 'shape'
```

Edoardo-Pedicillo commented on src/qibocal/calibrations/characterization/ro_optimization.py:
```diff
@@ -0,0 +1,576 @@
+    )
+
+    # execute the first pulse sequence
+    state0_results = platform.sweep(
+        state0_sequence, frequency_sweeper, nshots=nshots, average=False
+    )
+
+    for qubit in qubits:
+        r = {
+            k: v.ravel() for k, v in state0_results[ro_pulses[qubit].serial].raw.items()
+        }
+        r.update(
+            {
+                "frequency[Hz]": np.repeat(
+                    np.vstack(delta_frequency_range).T,
+                    len(np.arange(nshots)),
+                    axis=0,
+                ).flatten()
+                + ro_pulses[qubit].frequency,
+                "delta_frequency[Hz]": np.repeat(
+                    np.vstack(delta_frequency_range).T,
+                    len(np.arange(nshots)),
+                    axis=0,
+                ).flatten(),
+                "qubit": [qubit] * nshots * len(delta_frequency_range),
+                "iteration": np.repeat(
+                    np.vstack(np.arange(nshots)).T,
+                    len(delta_frequency_range),
+                    axis=1,
+                ).flatten(),  # Might be the other way depending on how is result happening. Axis=0 gives 123123 and axis=1 gives 1112233
+                "state": [0] * nshots * len(delta_frequency_range),
+            }
+        )
+        data.add_data_from_dict(r)
+    yield data
```
Here, the same code is repeated for state 0 and 1, I suggest defining a specific function.

Edoardo-Pedicillo commented on src/qibocal/calibrations/characterization/ro_optimization.py:
```diff
@@ -0,0 +1,576 @@
+    )
+
+    # execute the first pulse sequence
+    state0_results = platform.sweep(
+        state0_sequence, amplitude_sweeper, nshots=nshots, average=False
+    )
+
+    for qubit in qubits:
+        r = {
+            k: v.ravel() for k, v in state0_results[ro_pulses[qubit].serial].raw.items()
+        }
+        r.update(
+            {
+                "amplitude[dimensionless]": np.repeat(
+                    np.vstack(delta_amplitude_range).T,
+                    len(np.arange(nshots)),
+                    axis=0,
+                ).flatten()
+                * ro_pulses[qubit].amplitude,
+                "delta_amplitude[dimensionless]": np.repeat(
+                    np.vstack(delta_amplitude_range).T,
+                    len(np.arange(nshots)),
+                    axis=0,
+                ).flatten(),
+                "qubit": [qubit] * nshots * len(delta_amplitude_range),
+                "iteration": np.repeat(
+                    np.vstack(np.arange(nshots)).T,
+                    len(delta_amplitude_range),
+                    axis=1,
+                ).flatten(),  # Might be the other way depending on how is result happening. Axis=0 gives 123123 and axis=1 gives 1112233
+                "state": [0] * nshots * len(delta_amplitude_range),
+            }
+        )
+        data.add_data_from_dict(r)
+    yield data
```
it seems you can use the same function as above

Edoardo-Pedicillo commented on src/qibocal/calibrations/characterization/ro_optimization.py:
```diff
@@ -0,0 +1,576 @@
+    yield data
+
+    # execute the second pulse sequence
+    state1_results = platform.sweep(
+        state1_sequence, amplitude_sweeper, nshots=nshots, average=False
+    )
+
+    # retrieve and store the results for every qubit)
+    for qubit in qubits:
+        r = {
+            k: v.ravel() for k, v in state1_results[ro_pulses[qubit].serial].raw.items()
+        }
+        r.update(
+            {
+                "amplitude[dimensionless]": np.repeat(
+                    np.vstack(delta_amplitude_range).T,
+                    len(np.arange(nshots)),
+                    axis=0,
+                ).flatten()
+                * ro_pulses[qubit].amplitude,
+                "delta_amplitude[dimensionless]": np.repeat(
+                    np.vstack(delta_amplitude_range).T,
+                    len(np.arange(nshots)),
+                    axis=0,
+                ).flatten(),
+                "qubit": [qubit] * nshots * len(delta_amplitude_range),
+                "iteration": np.repeat(
+                    np.vstack(np.arange(nshots)).T,
+                    len(delta_amplitude_range),
+                    axis=1,
+                ).flatten(),  # Might be the other way depending on how is result happening. Axis=0 gives 123123 and axis=1 gives 1112233
+                "state": [1] * nshots * len(delta_amplitude_range),
+            }
+        )
+        data.add_data_from_dict(r)
+
+    # finally, save the remaining data and the fits
+    yield data
```
also here

Edoardo-Pedicillo commented on src/qibocal/calibrations/characterization/ro_optimization.py:
```diff
@@ -0,0 +1,576 @@
+        for qubit in qubits:
+            platform.set_lo_twpa_frequency(qubit, initial_frequency[qubit] + frequency)
+
+        state0_results = platform.execute_pulse_sequence(state0_sequence, nshots=nshots)
+        for qubit in qubits:
+            r = state0_results[ro_pulses[qubit].serial].raw
+            r.update(
+                {
+                    "frequency[Hz]": [platform.get_lo_twpa_frequency(qubit)] * nshots,
+                    "delta_frequency[Hz]": [frequency] * nshots,
+                    "qubit": [qubit] * nshots,
+                    "iteration": np.arange(nshots),
+                    "state": [0] * nshots,
+                }
+            )
+            data.add_data_from_dict(r)
```
Also here you can define a function and using it for states 0/1, I will not repeat the same for all the code, I think the concept is clear :)

Edoardo-Pedicillo commented on src/qibocal/plots/ro_optimization.py:
```diff
@@ -0,0 +1,526 @@
+
+    report_n += 1
+
+    # Show data for the first amplitude
+    fig.data[0].visible = True
+    fig.data[1].visible = True
+    fig.data[2].visible = True
+    fig.data[3].visible = True
```
you can use a `for` cycle

Edoardo-Pedicillo commented on src/qibocal/fitting/methods.py:
```diff
@@ -1073,3 +1075,118 @@ def calibrate_qubit_states_fit(data, x, y, nshots, qubits, degree=True):
+    )
+
+    # Use np.take_along_axis to get the correct indices for the threshold calculation
+    threshold = (
+        np.take_along_axis(iq_complex_sorted[0, ...], argmax, axis=-1)
+        + np.take_along_axis(iq_complex_sorted[1, ...], argmax, axis=-1)
+    ) / 2
+
```
I don't understand why the `threshold` is evaluated in this way, could you explain it ?

maxhant commented:
@aorgazf 
I've implemented your changes. It is the same but code differently. If you want to verify that it gives the same results. You can use the following:
```python
from qibocal.data import DataUnits
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import pathlib
from qibocal.fitting.methods import ro_optimization_fit, pint_to_float
from qibocal.plots import ro_frequency
from qibocal.data import Data
from plotly.subplots import make_subplots
import plotly.graph_objects as go

def ro_optimization_fit_alvaro(data, *labels, debug=False):
    """
    Fit the fidelities from parameters swept as labels, and extract rotation angle and threshold
    Args:
        data (Data): data to fit
        labels (str): variable used in the routine with format "variable_name"
    Returns:
        Data: data with the fit results
    """
    quantities = [
        *labels,
        "rotation_angle",
        "threshold",
        "fidelity",
        "assignment_fidelity",
        "average_state0",
        "average_state1",
    ]
    data_fit = Data(
        name="fit",
        quantities=quantities,
    )

    # Create a ndarray of i + 1j * q complex values with shape (qubit, parameter, state, shot)
    num_shots = len(data.df["iteration"].unique())
    shape = (*[len(data.df[label].unique()) for label in labels],)
    iq_complex = (
        data.df["i"].pint.magnitude.to_numpy()
        + 1j * data.df["q"].pint.magnitude.to_numpy()
    ).reshape(shape)

    # Rearrange axes
    labels = list(labels)
    iq_complex = np.moveaxis(
        iq_complex, [labels.index("state"), labels.index("iteration")], [-2, -1]
    )
    shape = iq_complex.shape

    # Calculate mean ground and excited states
    iq_mean_state0 = np.mean(np.take(iq_complex, 0, axis=-2), axis=-1, keepdims=True)
    iq_mean_state1 = np.mean(np.take(iq_complex, 1, axis=-2), axis=-1, keepdims=True)

    # Calculate rotation angle
    rotation_angle = np.expand_dims(np.angle(iq_mean_state1 - iq_mean_state0), axis=-2)

    # Rotate the data so that the mean of both states have the same imaginary component
    iq_complex_rotated = iq_complex * np.exp(-1j * rotation_angle)

    def calculate_cumulative_distributions(real_values_combined):
        """
        Calculates the cumulative distributions for state 0 and state 1
        Args:
            real_values_combined (np.ndarray): a flattened array of size (2 x num_shots)
        """
        sorted_real_values_state = np.sort(np.split(real_values_combined, 2))
        sorted_real_values_combined = np.sort(real_values_combined)

        cumulative_distributions = [[], []]
        for state in [0, 1]:
            app = 0
            for val in sorted_real_values_combined:
                app += np.max(
                    np.searchsorted(sorted_real_values_state[state][app:], val), 0
                )
                cumulative_distributions[state].append(app)
        return np.array(cumulative_distributions)

    # Select the real components and combine the data of both states
    iq_real = iq_complex_rotated.copy().real.reshape(*shape[:-2], 2 * num_shots)

    # Calculate the cumulative distributions
    cumulative_distributions = (
        np.apply_along_axis(calculate_cumulative_distributions, axis=-1, arr=iq_real)
        / num_shots
    )
    cumulative_distribution_0 = np.take(cumulative_distributions, 0, axis=-2)
    cumulative_distribution_1 = np.take(cumulative_distributions, 1, axis=-2)
    # Find the index of the threshold for which the difference between the cumulative distributions
    # of the two states is maximum
    argmax = np.argmax(
        np.abs(cumulative_distribution_0 - cumulative_distribution_1),
        axis=-1,
        keepdims=True,
    )

    # Get threshold value
    threshold = np.take_along_axis(iq_real, argmax, axis=-1)

    # Calculate the fidelity
    fidelity = np.take_along_axis(
        np.abs(cumulative_distribution_0 - cumulative_distribution_1), argmax, axis=-1
    )

    assignment_fidelity = (
        1
        - (
            (1 - np.take_along_axis(cumulative_distribution_0, argmax, axis=-1))
            + np.take_along_axis(cumulative_distribution_1, argmax, axis=-1)
        )
        / 2
    )

    # Add all the results to the data with labels as subnet without "state", "iteration"
    data_fit.df = (
        data.df.drop_duplicates(
            subset=[i for i in labels if i not in ["state", "iteration"]]
        )
        .reset_index(drop=True)
        .apply(pint_to_float)
    )
    data_fit.df["rotation_angle"] = rotation_angle.flatten()
    data_fit.df["threshold"] = threshold.flatten()
    data_fit.df["fidelity"] = fidelity.flatten()
    data_fit.df["assignment_fidelity"] = assignment_fidelity.flatten()
    data_fit.df["average_state0"] = iq_mean_state0.flatten()
    data_fit.df["average_state1"] = iq_mean_state1.flatten()

    if debug:
        return data_fit, cumulative_distribution_0 + 1j * cumulative_distribution_1, iq_complex
    else:
        return data_fit


folder = "debug_2"
routine = "ro_frequency"
param = "delta_frequency"
labels = ["state", "qubit", "iteration", param]
format = "csv"
name = "data"
qubits = [0,1,2]
data = DataUnits.load_data(pathlib.Path(__file__).parent / folder, "data", routine, format, name)

fit_compare = {"alvaro":
                   {"function": ro_optimization_fit_alvaro,
                        },
                "default":
                    {"function": ro_optimization_fit,
                        },
                }

# Compare the two methods by plotting the results for all qubits
fig = make_subplots(rows=len(qubits), cols=1, shared_xaxes=True)

for method in ["default", "alvaro"]:
    data_fit, cum_dist, iq_complex = fit_compare[method]["function"](data, *labels, debug=True)
    for i, qubit in enumerate(qubits): 
        dt = data_fit.df[data_fit.df["qubit"] == qubit]
        fig.add_trace(
            go.Scatter(
                x=dt[param],
                y=dt["fidelity"],
                name=f"{method} qubit {qubit}",
            ),
            row=i + 1,
            col=1,
        )          
fig.show()
fig.write_image(pathlib.Path(__file__).parent / folder / f"{routine}_fidelity_same.png")
```

maxhant commented:
I still to run it, and add the assignement_fidelity instead in the plotting

andrea-pasquale reviewed (changes requested):
Thanks @maxhant.
I tested the routines and they are working.
You can find below some comments. They are minor changes to simplify the code and some errors regarding documenation.

Are you sure that the fidelity is computed correctly, now?
![image](https://user-images.githubusercontent.com/49183315/235601132-5a716b05-fb53-4df3-bd7b-1d6568f6e438.png)
It seems a bit low to me.

maxhant commented on src/qibocal/calibrations/characterization/ro_optimization.py:
```diff
@@ -0,0 +1,483 @@
+                        np.vstack(np.arange(nshots)).T,
+                        len(delta_frequency_range),
+                        axis=1,
+                    ).flatten(),  # Might be the other way depending on how is result happening. Axis=0 gives 123123 and axis=1 gives 1112233
```
Iteration is the shot number, not the state

andrea-pasquale commented on src/qibocal/calibrations/characterization/ro_optimization.py:
```diff
@@ -0,0 +1,483 @@
+from qibolab.sweeper import Parameter, Sweeper
+
+from qibocal import plots
+from qibocal.data import Data, DataUnits
```
`Data` not used

andrea-pasquale commented on src/qibocal/calibrations/characterization/ro_optimization.py:
```diff
@@ -0,0 +1,483 @@
+
+    Args:
+        platform (:class:`qibolab.platforms.abstract.AbstractPlatform`): custom abstract platform on which we perform the calibration.
+        qubits (list): List of target qubits to perform the action
```
```suggestion
        qubits (dict): Dict of target qubits to perform the action
```

andrea-pasquale reviewed (commented):
Thanks for the updates @maxhant.
A few minor corrections.

Edoardo-Pedicillo reviewed (approved):
Thank @maxhant, LGTM
